### ISSUE TITLE
Allow yaml anchors in docker-compose

### ISSUE THREAD

**Issue by WolfspiritM** [microsoft/DockerTools#105]
Hi,

I'm trying to make a big compose file a bit more readable by combining globally used environment variables and labels into extension fields: https://docs.docker.com/compose/compose-file/#extension-fields.

```yml
version: '3.4'

x-globalenvs: &globalenvs
  GlobalEnv: "Test"

services:
  test:
    image: ${DOCKER_REGISTRY}test
    build:
      context: .
      dockerfile: src/Test/Dockerfile   
    environment:
      <<: *globalenvs
```

This breaks the Visual Studio docker integration:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error		(Line: 13, Col: 11, Idx: 223) - (Line: 13, Col: 22, Idx: 234): Anchor 'globalenvs' not found	docker-compose	C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Sdks\Microsoft.Docker.Sdk\build\Microsoft.VisualStudio.Docker.Compose.targets	294	
```

Also using anchors for other parts of the docker-compose fails in some cases aswell for example this tells me labels-anchor is not found while writing the labels into the label tag and referencing the anchor in deploy.labels works without issues:
```yml
version: '3.4'

services:
  test:
    image: ${DOCKER_REGISTRY}test
    build:
      context: .
      dockerfile: src/Test/Dockerfile
    deploy:
      labels: &label-anchor
        test: test
    labels:
      <<: *label-anchor
```

I also made an example repository.
It works fine when using ```docker-compose up``` but breaks when using Visual Studio:
https://github.com/WolfspiritM/DockerAnchorTest

I really hope this can be supported...

**Comment by dmikov** [comment]
Any updates on that? This is most basic functionality of yml, not even docker.

**Comment by AlexHaigh** [comment]
Any updates for 2021?

**Comment by AlexHaigh** [comment]
@NCarlsonMSFT I did some further testing and it works with the linked project.

What doesn't work is when you refence the anchor without the merge "<<" to use it without overriding a value.

Take the following example, which works in my Azure CI pipelines and with docker-compose. But does not work with Visual Studio.

`version: "3.5"

x-http: &default-test
  image: "${SERVICE_IMAGE_NAME}:${VERSION:-1}"
  build:
    context: .
    dockerfile: src/Test/Dockerfile
    args:
      SOLUTION: Test.sln
      PROJECT: src/Test
      ASSEMBLYNAME: Test
      
services:
  test: # This should use the extension as it is (does not work in VS, works everywhere else)
    *default-test
  second_test:
    << : *default-test #This should override (works in VS)
    image: "${SERVICE_IMAGE_NAME}:latest"`

**Comment by NCarlsonMSFT** [comment]
@AlexHaigh I did a bit more digging, and this error is coming from https://github.com/aaubry/YamlDotNet which we leverage to do our parsing. I'm not sure why, but in this case it rejects referencing the anchor without using the merge key.
As a work-around you can use the << : *default-test syntax even when you aren't going to override.
e.g.:
```
version: "3.5"

x-http: &default-test
  image: "${SERVICE_IMAGE_NAME}:${VERSION:-1}"
  build:
    context: .
    dockerfile: src/Test/Dockerfile
    args:
      SOLUTION: Test.sln
      PROJECT: src/Test
      ASSEMBLYNAME: Test
      
services:
  test:
    << : *default-test
  second_test:
    << : *default-test
    image: "${SERVICE_IMAGE_NAME}:latest"
```

**Comment by Spaier** [comment]
VS 2022 17.5.3

Even `<< : *default-test` stopped working

**Comment by dbreshears** [comment]
@Spaier,   this should fixed in the next 17.6 Preview slated for sometime next week.
